### DISCUSSION TITLE
Github Actions args not honoring variables

### DISCUSSION THREAD

**Discussion by thedarkwriter** [community/community#26126] [Actions]
<p>Hello!<br>
I have an action which runs a script that was built in a container from a previous step. I’m trying to execute that script and use a mix of envs, secrets, and args, but it seems like they are not being interpolated (or I have it configured incorrectly).</p>
<p>Here is the action:</p>
<pre><code>action "Script Runner" {
  needs = ["Docker Build", "Integration Branch"]
  uses = "docker://us.gcr.io/cool-project/cool-container:integration"
  env = {
    PROJECT = "project1"
    SERVICE = "app1"
    ROLE = "my_encrypt_role"
  }
  secrets = ["PASSWORD", "TOKEN"]
  runs = "local_scripts/deploy.sh"
  args = ["$PROJECT", "$SERVICE", "$TOKEN", "$ROLE", "$PASSWORD"]
}
</code></pre>
<p>Here is the logged event:</p>
<pre><code>### STARTED Script Runner 15:04:11Z

Already have image (with digest): redacted
Can't open /root/$ROLE for reading, No such file or directory ##YOU CAN SEE ROLE WASN"T READ
139722621944256:error:02001002:system library:fopen:No such file or directory:../crypto/bio/bss_file.c:72:fopen('/root/$ROLE','r')
139722621944256:error:2006D080:BIO routines:BIO_new_file:no such file:../crypto/bio/bss_file.c:79:
You are in project:=============&gt;$PROJECT . ##PROJECT WASN'T READ
You are deploying to service:===&gt;$SERVICE . ## SERVICE WASN'T READ

### FAILED Script Runner 15:04:12Z (486ms)
</code></pre>
<p>I’ve marked it, but you can see $ROLE, $PROJECT, and $SERVICE were not echoed (by the script) as their interpolated values but as literals. I’d expect this if I wrapped the args in single quotes. What am I doing wrong here?</p>
<p>Thank you! </p>

**Comment by wei** [comment]
<p>Args can be read out as $1, $2, $3, $4, $5 inside deploy.sh.</p>
<p>That being said, Environment variables should be accessible the way you’re using them, you’ll have to review Dockerfile and deploy.sh to identify the problem. You can start by specifying </p>
<p>runs = “env”</p>
<p>with no arg to look at all the environment variables present.</p>

**Comment by thedarkwriter** [comment]
<p>Hi <a class="mention" href="/wei">@wei</a> </p>
<p>I would think that is how it works, but the very top of the script reads in passed args.</p>
<p>PROJECT=$1<br>
shift<br>
SERVICE=$1<br>
shift</p>
<p>…</p>
<p>The project would be $1 and it does not post correctly. However, if I put in my args like this</p>
<p>args = [“project1”, “app1”, “TOKEN”, “role1”, “PASSWORD”] (PASSWORD and TOKEN are secrets)</p>
<p>Then my log comes back better:</p>
<pre><code>bad decrypt

You are in project:=============&gt;project1

You are deploying to service:===&gt;app1
</code></pre>
<p> But I can tell the secrets still didn’t make it through correctly, and I’d rather not expose it to debug but I may have to.</p>

**Comment by wei** [comment]
<p>Just create a new test secret and see if it comes back. Regardless, secrets are masked with [FILTERED] in the output anyways. Try my “env” test mentioned above if you still have troubles.</p>

**Comment by thedarkwriter** [comment]
<p>The runs env block, are you suggesting a block right before the script block? Such as</p>
<p>action “environment check” {<br>
  needs = [“Previous step”]<br>
  uses = “env”<br>
}</p>
<p>Also, I echoed out all passed args to the script then cat’d that script. I also wanted to see if there is a difference from my test secret being with our without $.</p>
<p>My result is, they were passed but not interpolated.</p>
<p>My test secret was TESTSECRET with a value of: TESTTESTTEST</p>
<p>Args looked like this:</p>
<p>args = “projec1 app1 $TOKEN encrypted_role $PASS TESTSECRET $TESTSECRET”</p>
<p>My cat (from the deploy.sh script) showed this:</p>
<p>project1 app1 $TOKEN encrypted_role $PASS TESTSECRET $TESTSECRET</p>

**Comment by wei** [comment]
<p>Try echo out env in the deploy.sh using echo $(env) to make sure it contains everything you want.</p>
<p>And you have to have $ before variables otherwise they are not interpolated.</p>

**Comment by thedarkwriter** [comment, marked as answer]
<p>Okay, I solved this. It looks like secrets are not passed at all through arguments. They are <strong>only</strong> exposed as environment variables. So to use my secrets, I had to slurp them in the bash script not pass them as arguments to the script from actions. For those who come after me and beat their head against the wall, this is what I did per secret</p>
<p>#!/bin/bash</p>
<p>…</p>
<p>PASS=<code>printf '%s\n' "${PASS}"</code><br>
SLACK_TOKEN=<code>printf '%s\n' "${TOKEN}"</code></p>
<p>…</p>
<p>doing_something_username password $PASS</p>
<p>post_in_slack $TOKEN</p>
